src/managers/MerkleProofManager.sol

```diff
@@ -0,0 +1,137 @@
+        require(strategistPolicy != bytes32(0), NotAStrategist());
+
+        for (uint256 i; i < calls.length; ++i) {
+            bytes memory addresses = abi.decode(_staticCall(calls[i].decoder, calls[i].targetData), (bytes));
+            PolicyLeaf memory leaf = _toPolicyLeaf(calls[i], addresses);
+
+            require(
+                MerkleProofLib.verify(calls[i].proof, strategistPolicy, _computeHash(leaf)),
+                InvalidProof(leaf, calls[i].proof)
+            );
+
+            emit ExecuteCall(calls[i].target, bytes4(calls[i].targetData), calls[i].targetData, calls[i].value);
+            _callWithValue(calls[i].target, calls[i].targetData, calls[i].value);
+        }
+    }
+
+    //----------------------------------------------------------------------------------------------
+    // Helpers
+    //----------------------------------------------------------------------------------------------
+
+    /// @dev Convert call data + decoded addresses into a merkle tree leaf
+    function _toPolicyLeaf(Call memory call, bytes memory addresses) internal pure returns (PolicyLeaf memory) {
+        return PolicyLeaf({
+            decoder: call.decoder,
+            target: call.target,
+            selector: bytes4(call.targetData),
+            addresses: addresses,
+            valueNonZero: call.value > 0
+        });
+    }
+
+    /// @dev Convert hash of a merkle tree leaf
+    function _computeHash(PolicyLeaf memory leaf) internal pure returns (bytes32) {
+        return keccak256(abi.encodePacked(leaf.decoder, leaf.target, leaf.valueNonZero, leaf.selector, leaf.addresses));
+    }
+
+    /// @dev Execute a static call to a contract
+    function _staticCall(address target, bytes memory data) internal view returns (bytes memory) {
+        (bool success, bytes memory returnData) = target.staticcall(data);
+        require(success, CallFailed());
+
+        return returnData;
+    }
+
+    /// @dev Execute a call with value to a contract
+    function _callWithValue(address target, bytes memory data, uint256 value) internal returns (bytes memory) {
+        require(address(this).balance >= value, InsufficientBalance());
+
+        (bool success, bytes memory returnData) = target.call{value: value}(data);
+        require(success, WrappedError(target, bytes4(data), returnData, abi.encodeWithSelector(CallFailed.selector)));
+
+        return returnData;
+    }
+}
+
+contract MerkleProofManagerFactory is IMerkleProofManagerFactory {
+    address public immutable spoke;
+    IBalanceSheet public immutable balanceSheet;
+
+    constructor(address spoke_, IBalanceSheet balanceSheet_) {
+        spoke = spoke_;
+        balanceSheet = balanceSheet_;
+    }
+
+    /// @inheritdoc IMerkleProofManagerFactory
+    function newManager(PoolId poolId) external returns (IMerkleProofManager) {
+        MerkleProofManager manager =
+            new MerkleProofManager{salt: bytes32(uint256(poolId.raw()))}(poolId, spoke, balanceSheet);
```

Comment:
What if a malicious actor creates this manager before the correct manager one? I think the correct would never can create its manager.

I think this can not be permissionless if it's per pool, or we should add some `salt` as a parameter.

Reply:
Ok, now I see your point.